packages/flutter/lib/src/painting/fractional_offset.dart

```diff
@@ -8,6 +8,191 @@ import 'package:flutter/foundation.dart';
+  ///
+  /// If [lerp] is applied to two objects of the same type ([FractionalOffset] or
+  /// [FractionalOffsetDirectional]), an object of that type will be returned (though
+  /// this is not reflected in the type system). Otherwise, an object
+  /// representing a combination of both is returned. That object can be turned
+  /// into a concrete [FractionalOffset] using [resolve].
+  static FractionalOffsetGeometry lerp(FractionalOffsetGeometry a, FractionalOffsetGeometry b, double t) {
+    if (a == null && b == null)
+      return null;
+    if ((a == null || a is FractionalOffset) && (b == null || b is FractionalOffset))
+      return FractionalOffset.lerp(a, b, t);
+    if ((a == null || a is FractionalOffsetDirectional) && (b == null || b is FractionalOffsetDirectional))
+      return FractionalOffsetDirectional.lerp(a, b, t);
+    if (a == null) {
+      return new _MixedFractionalOffset(
+        ui.lerpDouble(0.5, b._dx, t),
+        ui.lerpDouble(0.0, b._start, t),
+        ui.lerpDouble(0.5, b._dy, t),
+      );
+    }
+    if (b == null) {
+      return new _MixedFractionalOffset(
+        ui.lerpDouble(a._dx, 0.5, t),
+        ui.lerpDouble(a._start, 0.0, t),
+        ui.lerpDouble(a._dy, 0.5, t),
+      );
+    }
+    return new _MixedFractionalOffset(
+      ui.lerpDouble(a._dx, b._dx, t),
+      ui.lerpDouble(a._start, b._start, t),
+      ui.lerpDouble(a._dy, b._dy, t),
+    );
+  }
+
+  /// Convert this instance into a [FractionalOffset], which uses literal
+  /// coordinates (the `x` coordinate being explicitly a distance from the
+  /// left).
+  ///
+  /// See also:
+  ///
+  ///  * [FractionalOffset], for which this is a no-op (returns itself).
+  ///  * [FractionalOffsetDirectional], which flips the horizontal direction
+  ///    based on the `direction` argument.
+  FractionalOffset resolve(TextDirection direction);
+
+  @override
+  String toString() {
+    if (_start == 0.0) {
+      if (_dx == 0.0 && _dy == 0.0)
+        return 'FractionalOffset.topLeft';
+      if (_dx == 0.5 && _dy == 0.0)
+        return 'FractionalOffset.topCenter';
+      if (_dx == 1.0 && _dy == 0.0)
+        return 'FractionalOffset.topRight';
+      if (_dx == 0.0 && _dy == 0.5)
+        return 'FractionalOffset.centerLeft';
+      if (_dx == 0.5 && _dy == 0.5)
+        return 'FractionalOffset.center';
+      if (_dx == 1.0 && _dy == 0.5)
+        return 'FractionalOffset.centerRight';
+      if (_dx == 0.0 && _dy == 1.0)
+        return 'FractionalOffset.bottomLeft';
+      if (_dx == 0.5 && _dy == 1.0)
+        return 'FractionalOffset.bottomCenter';
+      if (_dx == 1.0 && _dy == 1.0)
+        return 'FractionalOffset.bottomRight';
+      return 'FractionalOffset(${_dx.toStringAsFixed(1)}, '
+                              '${_dy.toStringAsFixed(1)})';
+    }
+    if (_dx == 0.0) {
+      return 'FractionalOffsetDirectional(${_start.toStringAsFixed(1)}, '
+                                         '${_dy.toStringAsFixed(1)})';
```

Comment:
We should probably have named constants for the FractionalOffsetDirectional flavor too (e.g., FractionalOffsetDirectional.topStart)